README.md
# Introduction

This is a i18next cache layer to be used in the browser. It will load and cache resources from localStorage.

# Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/i18next-localstorage-cache), bower or [downloaded](https://github.com/i18next/i18next-localStorage-cache/blob/master/i18nextLocalStorageCache.min.js) from this repo.

- If you don't use a module loader it will be added to window.i18nextLocalstorageCache

```
# npm package
$ npm install i18next-localstorage-cache

# bower
$ bower install i18next-localstorage-cache
```

Wiring up:

```js
import i18next from 'i18next';
import Cache from 'i18next-localstorage-cache';

i18next
  .use(Cache)
  .init(i18nextOptions);
```

As with all modules you can either pass the constructor function (class) to the i18next.use or a concrete instance.

## Cache Options

**hint:** the cache plugin is disabled by default. Enable it by setting `enabled: true` in init options for production usage.

```js
{
  // turn on or off
  enabled: false,

  // prefix for stored languages
  prefix: 'i18next_res_',

  // expiration
  expirationTime: 7*24*60*60*1000,

  // language versions
  versions: {}
};
```

- Contrary to cookies behavior, the cache will respect updates to `expirationTime`. If you set 7 days and later update to 10 days, the cache will persist for 10 days

- Passing in a `versions` object (ex.: `versions: { en: 'v1.2', fr: 'v1.1' }`) will give you control over the cache based on translations version. This setting works along `expirationTime`, so a cached translation will still expire even though the version did not change. You can still set `expirationTime` far into the future to avoid this


Options can be passed in:

**preferred** - by setting options.cache in i18next.init:

```js
import i18next from 'i18next';
import Cache from 'i18next-localstorage-cache';

i18next
  .use(Cache)
  .init({
    cache: options
  });
```

on construction:

```js
  import Cache from 'i18next-localstorage-cache';
  const cache = new Cache(null, options);
```

via calling init:

```js
  import Cache from 'i18next-localstorage-cache';
  const cache = new Cache();
  cache.init(options);
```


CHANGELOG.md
### 1.1.1
- fixes i18nVersion is regarded as a namespace [PR7](https://github.com/i18next/i18next-localStorage-cache/pull/7)

### 1.1.0
- added "versions" to list of options, to allow control individually over cached languages

### 1.0.0

- fixes issue introduced with i18next v7.0.0 changes - if no lngs get passed to cache load we callback now to assert further processing


gulpfile.js
var gulp = require('gulp'),
    rename = require('gulp-rename'),
    babel = require('gulp-babel'),
    prompt = require('gulp-prompt'),
    git = require('gulp-git'),
    bump = require('gulp-bump'),
    filter = require('gulp-filter'),
    tag_version = require('gulp-tag-version'),
    shell = require('gulp-shell'),
    argv = require('yargs').argv,
    gulpif = require('gulp-if'),
    uglify = require('gulp-uglify'),
    sourcemaps = require('gulp-sourcemaps'),
    source = require('vinyl-source-stream'),
    buffer = require('vinyl-buffer'),
    browserify = require('browserify'),
    watchify = require('watchify'),
    babelify = require('babelify'),
    eslint = require('gulp-eslint'),
    Server = require('karma').Server;

var pkg = require('./package.json');

var entry = 'index.js',
    standaloneName = 'i18nextLocalStorageCache',
    output = 'index.js';

function compile(watch) {
  var bundler = browserify('./src/' + entry, { debug: argv.debug, standalone: standaloneName }).transform(babelify);
  if (watch) {
    bundle = watchify(bundler);
  }

  function rebundle() {
    bundler.bundle()
      .on('error', function(err) { console.error(err); this.emit('end'); })
      .pipe(source(output))
      .pipe(buffer())
      .pipe(gulpif(!argv.debug, uglify()))
      .pipe(gulpif(argv.debug, sourcemaps.init({ loadMaps: true })))
      .pipe(gulpif(argv.debug, sourcemaps.write('./')))
      .pipe(gulp.dest('./bin'));
  }

  if (watch) {
    bundler.on('update', function() {
      console.log('-> bundling...');
      rebundle();
    });
  }

  rebundle();
}

gulp.task('eslint', function () {
  return gulp.src(['src/**/*.js'])
    .pipe(eslint({
      useEslintrc: true
    }))
    .pipe(eslint.format())
    .pipe(eslint.failAfterError());
});

gulp.task('test', function (done) {
  new Server({
    configFile: __dirname + '/karma.conf.js',
    singleRun: true,
    reporters: [ 'spec', 'coverage' ],
  }, done).start();
});

gulp.task('test_compat', function (done) {
  new Server({
    configFile: __dirname + '/karma.backward.conf.js'
  }, done).start();
});


gulp.task('tdd', function (done) {
  new Server({
    configFile: __dirname + '/karma.conf.js'
  }, done).start();
});

gulp.task('babel', function () {
  return gulp.src('./src/**/*.js')
    .pipe(babel())
    .pipe(gulp.dest('./lib'));
});

gulp.task('rename', function () {
  return gulp
    .src('./bin/index.js')
    .pipe(rename('./' + standaloneName + '.min.js'))
    .pipe(gulp.dest('./'));
});

function inc(version) {
  if (!version) return;

  var type, tag;

  if (version.indexOf('.') < 0) {
    if (version === 'major' || version === 'minor' || version === 'patch') {
      type = version;
    } else {
      tag = version;

      var parts = pkg.version.split('-');
      if (parts.length > 1) {
        var tagVersion = 0;
        var p = parts[1].split('.');
        if (p[0] === tag) tagVersion = parseInt(p[1], 10) + 1;
        version = parts[0] + '-' + tag  + '.' + tagVersion;
      } else {
        version = pkg.version + '-' + tag  + '.0';
      }
    }
  }

  function cb(obj) {
    var stream = new require('stream').Transform({objectMode: true});
    stream._transform = function(file, unused, callback) {
      obj();
      callback(null, file);
    };
    return stream;
  }

  // get all the files to bump version in
  return gulp.src(['./package.json', './bower.json'])
    // bump the version number in those files
    .pipe(type ? bump({type: type}) : bump({version: version}))
    // save it back to filesystem
    .pipe(gulp.dest('./'))

    // commit change
    .pipe(git.commit('bumps package version'))

    // read only one file to get the version number
    .pipe(filter('package.json'))
    // **tag it in the repository**
    .pipe(tag_version({prefix: ''}))

    // push tag
    .pipe(prompt.confirm({
        message: 'Push tag ' + version + ' to github?',
        default: false
    }))
    .pipe(cb(function() {
      git.push('origin','master', {args: ' --tags'}, function (err) {
        if (err) throw err;
      });
    }))

    // npm publish
    .pipe(prompt.confirm({
        message: 'publish ' + version + ' to npm?',
        default: false
    }))
    .pipe(shell([
      'npm publish --tag ' + (tag ? tag : 'latest')
    ]));
}

function watch() {
  return compile(true);
};

gulp.task('concat', function() { return compile(); });
gulp.task('watch', function() { return watch(); });
gulp.task('bump', function() { return inc(argv.v); });

gulp.task('default', ['watch']);
gulp.task('build', ['concat', 'babel', 'rename']);
gulp.task('publish', ['bump']);


src/index.js
import * as utils from './utils';

const storage = {
  setItem(key, value) {
    if (window.localStorage) {
      try {
        window.localStorage.setItem(key, value);
      } catch (e) {
        // f.log('failed to set value for key "' + key + '" to localStorage.');
      }
    }
  },
  getItem(key, value) {
    if (window.localStorage) {
      try {
        return window.localStorage.getItem(key, value);
      } catch (e) {
        // f.log('failed to get value for key "' + key + '" from localStorage.');
      }
    }
    return undefined;
  }
};

function getDefaults() {
  return {
    enabled: false,
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

class Cache {
  constructor(services, options = {}) {
    this.init(services, options);

    this.type = 'cache';
    this.debouncedStore = utils.debounce(this.store, 10000);
  }

  init(services, options = {}) {
    this.services = services;
    this.options = utils.defaults(options, this.options || {}, getDefaults());
  }

  load(lngs, callback) {
    const store = {};
    const nowMS = new Date().getTime();

    if (!window.localStorage || !lngs.length) {
      return callback(null, null);
    }

    let todo = lngs.length;

    lngs.forEach((lng) => {
      let local = storage.getItem(this.options.prefix + lng);

      if (local) {
        local = JSON.parse(local);
        if (
          // expiration field is mandatory, and should not be expired
          local.i18nStamp && local.i18nStamp + this.options.expirationTime > nowMS &&

          // there should be no language version set, or if it is, it should match the one in translation
          this.options.versions[lng] === local.i18nVersion
        ) {
          delete local.i18nVersion;
          store[lng] = local;
        }
      }

      todo -= 1;
      if (todo === 0) {
        callback(null, store);
      }
    });
    return undefined;
  }

  store(storeParam) {
    const store = storeParam;
    if (window.localStorage) {
      for (const m in store) { // eslint-disable-line
        // timestamp
        store[m].i18nStamp = new Date().getTime();

        // language version (if set)
        if (this.options.versions[m]) {
          store[m].i18nVersion = this.options.versions[m];
        }

        // save
        storage.setItem(this.options.prefix + m, JSON.stringify(store[m]));
      }
    }
  }

  save(store) {
    this.debouncedStore(store);
  }
}

Cache.type = 'cache';

export default Cache;


rollup.config.js
import babel from 'rollup-plugin-babel';
import uglify from 'rollup-plugin-uglify';
import nodeResolve from 'rollup-plugin-node-resolve';
import { argv } from 'yargs';

const format = argv.format || argv.f || 'iife';
const compress = argv.uglify;

const babelOptions = {
  exclude: 'node_modules/**',
  presets: [['es2015', { modules: false }], 'stage-0'],
  babelrc: false
};

const dest = {
  amd: `dist/amd/i18nextLocalStorageCache${compress ? '.min' : ''}.js`,
  umd: `dist/umd/i18nextLocalStorageCache${compress ? '.min' : ''}.js`,
  iife: `dist/iife/i18nextLocalStorageCache${compress ? '.min' : ''}.js`
}[format];

export default {
  entry: 'src/index.js',
  format,
  plugins: [
    babel(babelOptions),
    nodeResolve({ jsnext: true })
  ].concat(compress ? uglify() : []),
  moduleName: 'i18nextLocalStorageCache',
  // moduleId: 'i18nextLocalStorageCache',
  dest
};


lib/LocalStorage.js
import * as utils from './utils';

let storage = {
  setItem: function(key, value) {
    if (window.localStorage) {
      try {
        window.localStorage.setItem(key, value);
      } catch (e) {
        //f.log('failed to set value for key "' + key + '" to localStorage.');
      }
    }
  },
  getItem: function(key, value) {
    if (window.localStorage) {
      try {
        return window.localStorage.getItem(key, value);
      } catch (e) {
        //f.log('failed to get value for key "' + key + '" from localStorage.');
        return undefined;
      }
    }
  }
};

function getDefaults() {
  return {
    enabled: false,
    prefix: 'i18next_res_',
    expirationTime: 7*24*60*60*1000
  };
}

class Cache {
  constructor(services, options = {}) {
    this.init(services, options);

    this.type = 'cache';
  }

  init(services, options = {}) {
    this.services = services;
    this.options = utils.defaults(options, this.options || {}, getDefaults());
  }

  load(lngs, callback) {
    let store = {}
      , nowMS = new Date().getTime();

    if(window.localStorage) {
      let todo = lngs.length;

      lngs.forEach(lng => {
        var local = storage.getItem(this.options.prefix + lng);

        if (local) {
          local = JSON.parse(local);
          if (local.i18nStamp && local.i18nStamp + this.options.expirationTime > nowMS) {
            store[lng] = local;
          }
        }

        todo--;
        if (todo === 0) callback(null, store);
      });
    }
  }

  save(store) {
    if(window.localStorage) {
      for (var m in store) {
        store[m].i18nStamp = new Date().getTime();
        storage.setItem(this.options.prefix + m, JSON.stringify(store[m]));
      }
    }
    return;
  }
}


export default Cache;


i18nextLocalStorageCache.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.i18nextLocalStorageCache = factory());
}(this, (function () { 'use strict';

var arr = [];
var each = arr.forEach;
var slice = arr.slice;

function defaults(obj) {
  each.call(slice.call(arguments, 1), function (source) {
    if (source) {
      for (var prop in source) {
        if (obj[prop] === undefined) obj[prop] = source[prop];
      }
    }
  });
  return obj;
}



function debounce(func, wait, immediate) {
  var timeout;
  return function () {
    var context = this,
        args = arguments;
    var later = function later() {
      timeout = null;
      if (!immediate) func.apply(context, args);
    };
    var callNow = immediate && !timeout;
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
    if (callNow) func.apply(context, args);
  };
}

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var storage = {
  setItem: function setItem(key, value) {
    if (window.localStorage) {
      try {
        window.localStorage.setItem(key, value);
      } catch (e) {
        // f.log('failed to set value for key "' + key + '" to localStorage.');
      }
    }
  },
  getItem: function getItem(key, value) {
    if (window.localStorage) {
      try {
        return window.localStorage.getItem(key, value);
      } catch (e) {
        // f.log('failed to get value for key "' + key + '" from localStorage.');
      }
    }
    return undefined;
  }
};

function getDefaults() {
  return {
    enabled: false,
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

var Cache = function () {
  function Cache(services) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Cache);

    this.init(services, options);

    this.type = 'cache';
    this.debouncedStore = debounce(this.store, 10000);
  }

  _createClass(Cache, [{
    key: 'init',
    value: function init(services) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      this.services = services;
      this.options = defaults(options, this.options || {}, getDefaults());
    }
  }, {
    key: 'load',
    value: function load(lngs, callback) {
      var _this = this;

      var store = {};
      var nowMS = new Date().getTime();

      if (!window.localStorage || !lngs.length) {
        return callback(null, null);
      }

      var todo = lngs.length;

      lngs.forEach(function (lng) {
        var local = storage.getItem(_this.options.prefix + lng);

        if (local) {
          local = JSON.parse(local);
          if (
          // expiration field is mandatory, and should not be expired
          local.i18nStamp && local.i18nStamp + _this.options.expirationTime > nowMS &&

          // there should be no language version set, or if it is, it should match the one in translation
          _this.options.versions[lng] === local.i18nVersion) {
            store[lng] = local;
          }
        }

        todo -= 1;
        if (todo === 0) {
          callback(null, store);
        }
      });
      return undefined;
    }
  }, {
    key: 'store',
    value: function store(storeParam) {
      var store = storeParam;
      if (window.localStorage) {
        for (var m in store) {
          // eslint-disable-line
          // timestamp
          store[m].i18nStamp = new Date().getTime();

          // language version (if set)
          if (this.options.versions[m]) {
            store[m].i18nVersion = this.options.versions[m];
          }

          // save
          storage.setItem(this.options.prefix + m, JSON.stringify(store[m]));
        }
      }
    }
  }, {
    key: 'save',
    value: function save(store) {
      this.debouncedStore(store);
    }
  }]);

  return Cache;
}();

Cache.type = 'cache';

return Cache;

})));
